### PR TITLE
Feature: customizable fields display for documents, saved views & dashboard widgets

### PR DESCRIPTION
**Pull request by shamoon**
## Proposed change

_This is a continuation of #6424 which importantly adds toggling fields etc to /documents page and for saved views (not just on the dashboard)_

This PR allows for customization of document display fields, saved views and dashboard widgets (screenshots below), including:

- Choosing display mode (table / small card / large card)
- Choose which fields to display for all display modes
- Settings per saved view (and dashboard widget) and for the documents list
- Including custom fields

Screenshot 2024-04-18 at 11 41 43 PM

Screenshot 2024-04-17 at 1 23 52 AM

Screenshot 2024-04-17 at 2 52 22 PM

Screenshot 2024-04-18 at 11 37 00 PM

Closes #4607
Closes #5568
Closes #4826
Kinda Closes #5298

## Type of change

- [ ] Bug fix: non-breaking change which fixes an issue.
- [x] New feature / Enhancement: non-breaking change which adds functionality. _Please read the important note above._
- [ ] Breaking change: fix or feature that would cause existing functionality to not work as expected.
- [ ] Documentation only.
- [ ] Other. Please explain:

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

**Comment by shamoon** [comment]
Went ahead and made the `display_fields` a simple JSON field with validation, the old way with dynamic choices was cool but it caused Django to try and create migrations for custom fields, Im good either way if you prefer the `DynamicOrderedMultipleChoiceField` but this seemed simpler.

**Comment by theomega** [comment]
Great job, I love this feature!
As promised here is my feedback:

- Hiding "Title" on list view make breaks the whole table layout. Suddenly the "tags" are in the "Owner" column. Maybe the easiest thing would be to not allow hiding the Title?
Screenshot 2024-04-21 at 15 28 25
- The "Show" dropdown contains "Notes" even if the Notes feature is disabled.
- I found the font-size in the "Show" dropdown inconsistent with the other dropdowns right next to it. All other dropdown have normal font sizes where as the "Show" dropdown has a smaller font size.
- When in "cards" display mode in the list (the middle button), then the behaviour of the "Added" and "Created" fields in the "Show" Dropdown is not super intuitive: They are both checked as default, when I unselect "Created", then nothing gets hidden, but the dates swap. It took my some digging in the code to understand that if I unselect "Created", but leave "Added" selected, then the "Added" date is chosen instead. Why not display really both dates if both are selected in the "Show" dropdown?
- I'm unsure if the "Share" field works as intentended in the "cards" display mode: Nothing changes if I enable or disable this field. I have shared a document (using the "Share Links" feature), but independent if I enable or disable "Share" in the dropdown, nothing displays. Same for the list mode. Maybe I'm misunderstand what "Share" should display.
- The "Show" dropdown does not respect permissions, so if the user does not have the right permissions to even see the data, we should hide the field in the dropdown
- One change that was in my (closed) PR was also hiding the "Owner" field in the "Sort" dropdown if the user does not have the permissions to even see the owners. Not sure if we should do this here in this PR or if I should open a separate one to `dev` for this.

**Comment by theomega** [comment]
One thing I'm completely lost about: Is this feature expected to work with saved views? So are the saved views expected to store which columns they display?
I can't get this to work:

If I reduce the columns displayed to only the "title" and then save the view, however I recall the view (side menu or via the "Views" dropdown), again all columns/fields are displayed. I think this is a saving problem, if I check the Django Admin for the Saved Views, they all have `null` inside their "Document display fields". So somehow the fields are not stored when saving the view.


Screenshot 2024-04-21 at 16 18 34

**Comment by shamoon** [comment]
Addressed those, thanks for testing.

The shared button shows the little note if a document is shared:
Screenshot 2024-04-21 at 8 16 54 AM
Screenshot 2024-04-21 at 8 16 57 AM

As far as the saved views goes, yes that absolutely should work, and I cant reproduce a problem with it. See screenshots with just title:

Screenshot 2024-04-21 at 7 48 04 AM
Screenshot 2024-04-21 at 7 46 21 AM
Screenshot 2024-04-21 at 7 46 15 AM
Screenshot 2024-04-21 at 8 14 28 AM

If no fields are specified it uses a default set (same as current)

**Comment by shamoon** [comment]
When creating the view with the POST request, the UI doesn't even provide the `display_fields` field in the request, this is where the `null` in the database comes from.

I cant repro that either:
Screenshot 2024-04-21 at 9 24 16 AM

**Comment by shamoon** [comment]
Shared is for permissions not share links. Table thing is fixed.

**Comment by theomega** [comment]
OK, then all good with the shared thing.

As for saving the columns:

I think this is the code that is supposed to construct what is sent to the backend when saving a view, correct?
https://github.com/paperless-ngx/paperless-ngx/blob/feature-dynamic-fields-display/src-ui/src/app/components/document-list/document-list.component.ts#L262

This doesn't contain the `display_fields` field at all. Am I looking and testing the right code?

**Comment by shamoon** [comment]
Ah, we were talking about two different things, you mean saving a new view, oh yea great catch.

**Comment by shamoon** [comment]
Alright, thats sorted too thanks 👍

**Comment by theomega** [comment]
Oh, I feel stupid now, I didn't know you can actually edit the Views in the "Settings". So I have never tested that :-D

I just spotted another bug: I just tried modifying an existing view in the display mode, but it doesn't even allow me to save.

How to reproduce:
- Save a new view
- Bug (as discussed): display_fields is not stored
- Now load the view (i.e. through the sidebar)
- Modify the displayed fields using the "Show" dropdown

Observe:
- You can't save/update the view, the only option is to create another view
- The "Save [viewname]" button is grayed out:
Screenshot 2024-04-21 at 18 54 28

Expected:
- I should be able to save/update the view with the new fields to show

**Comment by theomega** [comment]
Sorry, forget about that bug report, you just seem to have also fixed that....

**Comment by theomega** [comment]
Then this looks really awesome to me, nothing more from my side after testing this.

**Comment by tomihbk** [comment]
Dear @shamoon, many thanks for this amazing feature! Does this PR also include the possibility of sorting by custom fields ?

File: src/documents/serialisers.py [hunk 1]
```diff
@@ -810,12 +810,31 @@ class Meta:
...
             "permissions",
             "user_can_change",
             "set_permissions",
         ]
 
+    def validate(self, attrs):
+        attrs = super().validate(attrs)
+        if "display_fields" in attrs and attrs["display_fields"] is not None:
+            for field in attrs["display_fields"]:
+                if re.sub(r"\d+", "%d", field) == SavedView.DisplayFields.CUSTOM_FIELD:
```
Comment (stumpylog) [inline src/documents/serialisers.py:826]: What do you think about checking for `custom_field_ in field`, then the regex to match and extract, if possible?  Just thinking two regex checks here might be expensive?

Comment (shamoon) [reply src/documents/serialisers.py]: Sure, lmk what you think of this. We could probably even eliminate the regex altogether if you prefer (just remove the prefix and cast the rest to an int). Guess I just like regexs :)

**Review by stumpylog — APPROVED** [review]

**Pull request merged**